components/feature/session/src/main/java/mozilla/components/feature/session/SlowLoadingFeature.kt

```diff
@@ -0,0 +1,67 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.session
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.mapNotNull
+import mozilla.components.browser.state.selector.findTabOrCustomTabOrSelectedTab
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.concept.engine.script.SlowScriptCallback
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.support.base.feature.LifecycleAwareFeature
+import mozilla.components.support.base.log.logger.Logger
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+
+/**
+ * Feature implementation that provides a notification when the
+ * slowScript callback is called and enables the user to
+ * continue or stop running the script.
+ *
+ * This is where we'd observe state.content.slowLoadingScriptCallback.
+ * When it's null -> everything is okay, nothing to do
+ * When it's not null -> post a notification for the user to choose next actions:
+ *   - slowLoadingScriptCallback.allowSlowLoadingScript()
+ *   - slowLoadingScriptCallback.stopSlowLoadingScript()
+ *
+ * @param sessionId ID of the session running the slow script.
+ * @param store Reference to the application's [BrowserStore].
+ * @param showNotification Function having [SlowScriptCallback] as parameter used to show
+ * the notification with the controls for the callback.
+ * @param dismissNotification Function used to dismiss the notification.
+ */
+class SlowLoadingFeature(
+    private val sessionId: String?,
+    private val store: BrowserStore,
+    private val showNotification: (SlowScriptCallback) -> Unit,
+    private val dismissNotification: () -> Unit,
+) : LifecycleAwareFeature {
+
+    private val logger = Logger("SlowScript")
+
+    private var scope: CoroutineScope? = null
+
+    override fun start() {
+        logger.debug("SlowScriptFeature started.")
+        scope = store.flowScoped { flow ->
+            flow.mapNotNull { state -> state.findTabOrCustomTabOrSelectedTab(sessionId) }
+                .ifChanged { it.content.slowScriptCallback }
+                .collect {
+                    if (it.content.slowScriptCallback != null) {
+                        showNotification(it.content.slowScriptCallback!!)
+                    } else {
+                        dismissNotification()
+                    }
```

Comment:
Are we going to have different apps show different UIs? Normally on desktop we only show a dialog with two button continue or stop, I think we could have simple UI handle on AC if we need a custom UI per app we could add extra hooks later.

Reply:
I agree that we could use the UI being handled in AC, however we do not have any official specs regarding the dialog. I'll ask again to clarify this aspect.

Reply:
We could take a look to what we had in Fennec to draft some ideas from it :)

Reply:
> We could take a look to what we had in Fennec to draft some ideas from it :)

There was none. 🥲 

See [this link](https://searchfox.org/mozilla-esr68/source/mobile/android/app/mobile.js#363) that searches through the mozilla-central codebase during release 68 which is the last major release of fennec.

[`mozilla-esr68`](https://searchfox.org/mozilla-esr68/source/) is also listed on the searchfox.org homepage for easy reference.